File: system_simulations/network_model/Network.java
/* Composite */
public class Network implements AtomicModel{
	private AtomicModel[] components;
	private boolean state;
	private int componentCount;
	private int tick;

	private Network(){
	}

	public static NetworkBuilder builder(){
		return new NetworkBuilder();
	}

	public boolean lambda(){
		return state;
	}

	public void delta(boolean[] X){
		for (int i = 0; i < componentCount; i++){
			boolean[][] inputs = coupling(X);
			for (int j = 0; j < componentCount; j++){
				components[j].delta(inputs[j]); // input to the Network
			}
			System.out.println();
		}
		state = components[1].lambda();
		System.out.println("\nNetwork state: " + ((state) ? 1 : 0) + "\n");
		tick++;
	}

	private boolean[][] coupling(boolean[] X){
		boolean XOR1y = components[0].lambda(); // coupled with input to XOR2
		boolean XOR2y = components[1].lambda(); // coupled with output of network AND input to M
		boolean My = components[2].lambda(); // coupled with input to XOR2

		boolean[] XOR1x = X;
		boolean[] XOR2x = {XOR1y, My};
		boolean[] Mx = {My, XOR2y};
		boolean[][] retValue = {XOR1x, XOR2x, Mx};

		return retValue;
	}

	public int getTick(){
		return this.tick;
	}

	// coupling function push over to dest

	public static class NetworkBuilder {
		private AtomicModel[] components = new AtomicModel[3];
		private int componentCount = 0;

		public NetworkBuilder(){
		}

		public NetworkBuilder addComponent(AtomicModel component){
			this.components[componentCount] = component;
			this.componentCount++;
			return this;
		}

		public Network build(){
			Network network = new Network();
			network.components = this.components;
			network.componentCount = this.components.length;
			network.state = false;
			network.tick = 0;
			return network;
		}
	}
}

File: system_simulations/network_model/cpp/Network.cpp
#include <iostream>
#include "AtomicModel.h"
#include "Network.h"
using namespace std;

Network::Network(NetworkBuilder* builder){
	for (int i = 0; i < builder->compCount; i++){
		this->components[i] = builder->components[i];
	}
	this->compCount = builder->compCount;
	this->tick = 0;
}

bool Network::lambda(){
	return components[compCount-2]->lambda();
}

void Network::delta(bool x1, bool x2){
	for (int i = 0; i < compCount; i++){
		bool** inputs = couple(x1, x2);
		for (int j = 0; j < compCount; j++){
			cout << j << endl;
			components[j]->delta(inputs[j][0], inputs[j][1]);
		}
		cout << endl;
	}
	cout << "New Network Output: " << lambda() << endl;
}

bool** Network::couple(bool x1, bool x2){
	bool* outputs = new bool[compCount];
	for (int i = 0; i < compCount; i++){
		outputs[i] = components[i]->lambda();
	}

	bool* XOR1x = new bool[2]{x1, x2};
	bool* XOR2x = new bool[2]{outputs[0], outputs[2]};
	bool* Mx = new bool[2]{outputs[1], outputs[2]};
	//bool ret[compCount][2] = ;

	bool** retValue; //= new bool[3][2]{{XOR1x}, {XOR2x}, {Mx}};
	retValue[0] = XOR1x;
	retValue[1] = XOR2x;
	retValue[2] = Mx;

	return retValue;
}

Network::NetworkBuilder::NetworkBuilder(){
	this->compCount = 0;
}

Network::NetworkBuilder& Network::NetworkBuilder::addComponent(AtomicModel* component){
	this->components[this->compCount] = component;
	this->compCount++;
	return *this;
}

Network* Network::NetworkBuilder::build(){
	return new Network(this);
}